#10066 Update pydoctor version and remove Twisted pydoctor custom code that is now available as standard features.

adiroiban opened the pull request:
## Scope and purpose

Fixes #10066


If the epydoc API markup is invalid, the CI is happy and green.

We need to update the CI to fail on pydoctor build errors.
This should help make sure the API doc is valid and sane :)

As part of this PR I have updated the pydoctor version.

Also, instead of using the custom python script to call pydoctor, this calls pydoctor directly from tox.
I hope that in this way it should be easier to understand what is going on there, with fewer redirection layers.


## Contributor Checklist:

* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://docs.twistedmatrix.com/en/stable/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://github.com/twisted/trac-wiki-archive/blob/trunk//ReviewProcess.mediawiki#News_files))
* [x] The title of the PR starts with the associated issue number (without the `#` character).
* [x] I have updated the automated tests and checked that all checks for the PR are green.
* [x] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #123 from twisted/4356-branch-name-with-issue-id

Author: adiroiban
Reviewer: 
Fixes #10066

Update pydoctor version.
Fail CI for a PR on pydoctor build errors.
```

adiroiban committed:
Upgrade pydoctor. Create dedicated apidoc builder that fails on apidocs errors.

adiroiban committed:
Remove custom apidocs builder script and tests.

adiroiban committed:
Fix python setup.

adiroiban committed:
Run all docs in a separate job to make it easier to check the output.

tristanlatr commented on tox.ini:
```diff
@@ -158,10 +158,31 @@ commands =
 extras = dev_release
-commands = {toxinidir}/bin/admin/build-apidocs {toxinidir}/src/ apidocs
+commands =
+    pydoctor --quiet \
```
This is missing `--template-dir` option.

tristanlatr commented on setup.cfg:
```diff
@@ -56,7 +56,7 @@ test =
 dev_release =
     towncrier ~= 19.2
-    pydoctor ~= 21.9.0; python_version >= "3.6"
+    pydoctor ~= 22.5.0; python_version >= "3.6"
```
```suggestion
    pydoctor ~= 22.7.0; python_version >= "3.6"
```

tristanlatr replied:
Updating to pydoctor 22.7 means the explicit support for `twisted.python.deprecated` can be deleted. 
Almost everything in `twisted.python._pydoctor.TwistedSystem` can be deleted, since the privacy tweaks can be achieved with flag `--privacy`, there is only the `TwistedSphinxInventory` tweaks left.

adiroiban committed:
Merge branch 'trunk' into 10066-pydoctor-and-docs-ci

adiroiban committed:
Update after initial review.

adiroiban commented:
thanks @tristanlatr 


I have updated twisted/twisted-contributors team. Not sure why you weren't there.

If you have time. Can you make the required changes to get pydoctor working ?

Thanks

adiroiban commented:
> It’s like the _pydoctor module is not included in the default dist build.

I don't know why it is not included.
But I think that is a bug.

In the code, I see that the intention is to have it included.

https://github.com/twisted/twisted/blob/fc20692e9eb133d20d4c321534bd798836c2595e/MANIFEST.in#L40-L44

```
python setup.py bdist_wheel

adding 'twisted/python/_pydoctor.py'
```

and it was installed in my environment


-----------

tristanlatr committed:
Drop custom pydoctor system.

tristanlatr committed:
Hide tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

tristanlatr committed:
Configure privacy tweaks

tristanlatr committed:
Merge branch '10066-pydoctor-and-docs-ci' of github.com:twisted/twisted into 10066-pydoctor-and-docs-ci

tristanlatr committed:
Move all template customization in a one place

tristanlatr committed:
Fix typo

tristanlatr committed:
Fix paragraph indentation in _pydoctor.py

adiroiban committed:
Merge branch 'trunk' into 10066-pydoctor-and-docs-ci

adiroiban committed:
Fix ci.

adiroiban commented:
Ok, One test failing

```
Traceback (most recent call last):
  File "/home/runner/work/twisted/twisted/.tox/alldeps-withcov-posix/lib/python3.8/site-packages/twisted/trial/runner.py", line 688, in loadByName
    return self.suiteFactory([self.findByName(name, recurse=recurse)])
  File "/home/runner/work/twisted/twisted/.tox/alldeps-withcov-posix/lib/python3.8/site-packages/twisted/trial/runner.py", line 451, in findByName
    obj = reflect.namedModule(searchName)
  File "/home/runner/work/twisted/twisted/.tox/alldeps-withcov-posix/lib/python3.8/site-packages/twisted/python/reflect.py", line 156, in namedModule
    topLevel = __import__(name)
  File "/home/runner/work/twisted/twisted/.tox/alldeps-withcov-posix/lib/python3.8/site-packages/twisted/python/test/test_pydoctor.py", line 18, in <module>
    from twisted.python._pydoctor import TwistedSphinxInventory, TwistedSystem
  File "/home/runner/work/twisted/twisted/.tox/alldeps-withcov-posix/lib/python3.8/site-packages/twisted/python/_pydoctor.py", line 17, in <module>
    from pydoctor import astbuilder, model, zopeinterface  # type: ignore[import]
builtins.ImportError: cannot import name 'zopeinterface' from 'pydoctor' (/home/runner/work/twisted/twisted/.tox/alldeps-withcov-posix/lib/python3.8/site-packages/pydoctor/__init__.py)

twisted.python.test.test_pydoctor
```

adiroiban commented:
> I'll just cleanup _pydoctor.py is that OK with you.

Yes. Thanks.

Feel free to push anything.

I can then see how to break this PR to get these changes reviewed and merged.

tristanlatr committed:
Cleanup _pydoctor.py and test_pydoctor.py.

Newer versions of pydoctor includes support for twisted.python.deprecate and allows to tweak the privacy of some objects.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

tristanlatr committed:
Use config file not to duplicate pydoctor CLI args

adiroiban commented:
> The only thing that bothers me is that we’re not actually using the custom system, we just have tests for it.

We can remove the custom system code if not used.

less is more :)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

tristanlatr committed:
Completely drop custom pydoctor system.

tristanlatr committed:
Merge branch '10066-pydoctor-and-docs-ci' of github.com:twisted/twisted into 10066-pydoctor-and-docs-ci

tristanlatr committed:
Use readthedocs theme

glyph commented on docs/conf.py:
```diff
@@ -221,25 +221,16 @@
+    # pydoctor should not fail the sphinx build, we have another tox environment for that.
+    f"--config={_project_root}/setup.cfg",
     f"--html-viewsource-base=https://github.com/twisted/twisted/tree/{_git_reference}/src",
-    "--project-name=Twisted",
-    "--project-url=https://twistedmatrix.com/",
-    "--system-class=twisted.python._pydoctor.TwistedSystem",
-    "--docformat=epytext",
-    "--intersphinx=https://docs.python.org/3/objects.inv",
-    "--intersphinx=https://cryptography.io/en/latest/objects.inv",
-    "--intersphinx=https://pyopenssl.readthedocs.io/en/stable/objects.inv",
-    "--intersphinx=https://hyperlink.readthedocs.io/en/stable/objects.inv",
-    "--intersphinx=https://twisted.org/constantly/docs/objects.inv",
-    "--intersphinx=https://twisted.org/incremental/docs/objects.inv",
-    "--intersphinx=https://python-hyper.org/projects/hyper-h2/en/stable/objects.inv",
-    "--intersphinx=https://priority.readthedocs.io/en/stable/objects.inv",
-    "--intersphinx=https://zopeinterface.readthedocs.io/en/latest/objects.inv",
-    "--intersphinx=https://automat.readthedocs.io/en/latest/objects.inv",
```
yay for getting this to be declarative data in `setup.cfg` and not stuffed into command-line args!

tristanlatr replied:
Thanks

adiroiban commented:
These changes are awesome.

I love the config file.

-----------

The theme is also ok.

Just some feedback for the theme.

There are now 2 main links:

* Twisted  main site
* API docs index

![image](https://user-images.githubusercontent.com/204609/181659403-12814351-7442-4445-988e-e63bd4265c58.png)

Maybe instead of `Twisted  main site`  the link should be to the narrative docs base - like `https://twisted--11578.org.readthedocs.build/en/11578/index.html` this should match RTD links.


------------

Now, in order to have this merged ASAP, as you mentioned on the issue, I think that is best to reduce the scope of this PR.

Instead of pydoctor upgrade + ci fail, we should use this PR only to upgrade the pydoctor version.

I/we can look to fix the pydoctor errors in a separate PR.

glyph commented on bin/admin/build-apidocs:
```diff
@@ -1,19 +0,0 @@
-#!/usr/bin/env python
```
Love to see all the `./build/admin` scripts get replaced with tox environments. Much more discoverable and less error-prone.

glyph commented on setup.cfg:
```diff
@@ -214,3 +214,54 @@ per-file-ignores =
+docformat=epytext
+theme=readthedocs
+privacy=
+    HIDDEN:twisted.words.test
+    HIDDEN:twisted.web.test
+    HIDDEN:twisted.spread.test
+    HIDDEN:twisted.scripts.test
+    HIDDEN:twisted.runner.test
+    HIDDEN:twisted.python.test
+    HIDDEN:twisted.protocols.haproxy.test
+    HIDDEN:twisted.protocols.test
+    HIDDEN:twisted.positioning.test
+    HIDDEN:twisted.persisted.test
+    HIDDEN:twisted.pair.test
+    HIDDEN:twisted.names.test
+    HIDDEN:twisted.mail.test
+    HIDDEN:twisted.logger.test
+    HIDDEN:twisted.cred.test
+    HIDDEN:twisted.conch.test
+    HIDDEN:twisted.application.runner.test
+    HIDDEN:twisted.application.twist.test
+    HIDDEN:twisted.application.test
+    HIDDEN:twisted._threads.test
+    HIDDEN:twisted.trial._dist.test
+    HIDDEN:twisted.trial.test
+    HIDDEN:twisted.internet.test
+    HIDDEN:twisted.test.*
+    PUBLIC:twisted.test.proto_helpers
```
Beautiful. Self-documenting!

glyph reviewed (approved):
LGTM. Love to see all this unnecessary support code getting zapped.

glyph commented:
Auto-merge set so we can land this as soon as static check and api doc build issues are corrected. (Looks like maybe an intersphinx thing for the docs build?)

adiroiban committed:
Ignore pydoctor error for now. They will be fixed in #11590

adiroiban commented:
Ok. The plan is to land this PR asap.

The remaining pydoctor "compilation" error should be fixed as a followup in https://github.com/twisted/twisted/issues/11590

Those errors already exist in trunk, so they are not errors introduced in this PR.

Is just that with this PR, pydoctor is detecting more errors.
